File: src/main/java/com/epic/warpstones/storage/FileStorage.java
package com.epic.warpstones.storage;

import com.epic.warpstones.Warpstones;
import com.epic.warpstones.interfaces.WarpstoneStorage;
import com.epic.warpstones.models.Warpstone;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class FileStorage implements WarpstoneStorage {
    private JavaPlugin plugin;
    private File database;

    private final String DATABASE_FILE_NAME = "warpstones.dat";

    public FileStorage() throws IOException {
        this.plugin = JavaPlugin.getPlugin(Warpstones.class);

        File dataFolder = this.plugin.getDataFolder();

        if (!dataFolder.exists()) {
            dataFolder.mkdirs();
        }

        this.database =  new File(dataFolder, DATABASE_FILE_NAME);

        if (!this.database.exists()) {
            this.database.createNewFile();
        }
    }

    @Override
    public List<Warpstone> loadWarpstones() {
        List<Warpstone> warpstones = new ArrayList<>();

        try {
            List<String> lines = Files.readAllLines(Paths.get(this.database.getAbsolutePath()));

            for (String line : lines) {
                Warpstone warpstone = new Warpstone();
                String[] parts = line.split("\\|");

                warpstone.setX(Integer.parseInt(parts[0]));
                warpstone.setY(Integer.parseInt(parts[1]));
                warpstone.setZ(Integer.parseInt(parts[2]));
                warpstone.setName(parts[3]);
                warpstone.setDestination(parts[4]);
                warpstone.setOwner(UUID.fromString(parts[5]));
                warpstones.add(warpstone);
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        return warpstones;
    }

    @Override
    public void saveWarpstones(List<Warpstone> warpstones) {
        try {
            StringBuilder builder = new StringBuilder();

            for (int i = 0; i < warpstones.size(); i++) {
                Warpstone warpstone = warpstones.get(i);

                builder.append(warpstone.getX())
                        .append("|")
                        .append(warpstone.getY())
                        .append("|")
                        .append(warpstone.getZ())
                        .append("|")
                        .append(warpstone.getName())
                        .append("|")
                        .append(warpstone.getDestination())
                        .append("|")
                        .append(warpstone.getOwner().toString());

                if (i + 1 != warpstones.size()) {
                    builder.append("\n");
                }
            }

            Files.writeString(Paths.get(this.database.getAbsolutePath()), builder.toString());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/java/com/epic/warpstones/storage/FileStorage.kt
package com.epic.warpstones.storage

import com.epic.warpstones.Warpstones
import com.epic.warpstones.interfaces.WarpstoneStorage
import com.epic.warpstones.models.Warpstone
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.UUID

private const val DATABASE_FILE_NAME = "warpstones.dat"

class FileStorage: WarpstoneStorage {
    private val plugin: JavaPlugin
    private val database: File
    private val databaseAbsolutePath: Path

    constructor() {
        this.plugin = JavaPlugin.getPlugin(Warpstones::class.java)

        val dataFolder = this.plugin.dataFolder

        if (!dataFolder.exists()) {
            dataFolder.mkdirs()
        }

        this.database = File(dataFolder, DATABASE_FILE_NAME)

        if (!this.database.exists()) {
            this.database.createNewFile()
        }

        this.databaseAbsolutePath = Paths.get(this.database.absolutePath)
    }

    override fun loadWarpstones(): ArrayList<Warpstone> {
        val warpstones = ArrayList<Warpstone>()

        try {
            val lines = Files.readAllLines(this.databaseAbsolutePath)

            for (line in lines) {
                val warpstone = Warpstone()
                val parts = line.split('|')

                warpstone.x = parts[0].toDouble()
                warpstone.y = parts[1].toDouble()
                warpstone.z = parts[2].toDouble()
                warpstone.name = parts[3]
                warpstone.destination = parts[4]
                warpstone.owner = UUID.fromString(parts[5])
                warpstones.add(warpstone)
            }
        } catch (e: IOException) {
            throw RuntimeException("Could not load warpstones", e)
        }

        println("Loaded ${warpstones.size} warpstones")
        println(warpstones.toString())

        return warpstones
    }

    override fun saveWarpstones(warpstones: ArrayList<Warpstone>) {
        try {
            val builder = StringBuilder()

            for (warpstone in warpstones) {
                builder.append("${warpstone.x}|${warpstone.y}|${warpstone.z}|")
                builder.append("${warpstone.name}|${warpstone.destination}|${warpstone.owner.toString()}")
                builder.append("\n")
            }

            builder.trimEnd()

            Files.writeString(this.databaseAbsolutePath, builder.toString())
        } catch(e: IOException) {
            throw RuntimeException("Could not save warpstones", e)
        }
    }

}